torch.autograd.graph.increment_version: accept List[Tensor], use in AOTDispatcher

bdhirsh opened the pull request:
The regression from https://github.com/pytorch/pytorch/issues/132281 pinpoints https://github.com/pytorch/pytorch/commit/e4ace1a396265706ddd0041bb6a26c18d2de1095 as the cause. The main delta that commit introduces is that we now manually check `is_inference()` and call `increment_version()` (a pybind call) on every mutated input tensor to the graph.

This PR attempts to reduce overhead a bit by bundling up all of those checks into a single pybind call, by:

(1) updating `torch.autograd.graph.increment_version()` to accept a `Union[Tensor, List[Tensor]]`

(2) updating its semantics to no-op if you pass in a tensor with no version counter, instead of erroring
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132699
* __->__ #132652

bdhirsh committed:
torch.autograd.graph.increment_version: accept List[Tensor], use in AOTDispatcher

[ghstack-poisoned]

albanD commented on torch/autograd/graph.py:
```diff
@@ -227,8 +227,17 @@ def increment_version(tensor: torch.Tensor) -> None:
+    we will not bump its version counter (because your tensor does not have one).
+
+    Can accept either a tensor, or a list of tensors.
+    If you pass in a list of tensors, we will separately
```
You stopped mid sentence? :p 
No need for a paragraph for this, you can put it inline above.

bdhirsh replied:
... oops

albanD commented on torch/autograd/graph.py:
```diff
@@ -227,8 +227,17 @@ def increment_version(tensor: torch.Tensor) -> None:
+    if isinstance(tensor, list):
+        torch._C._increment_versions(tensor)
+    else:
+        torch._C._increment_version(tensor)
```
can we have only one binding for the sequence version and just wrap single Tensor into a one element tuple here?

bdhirsh replied:
sgtm!

albanD commented on torch/csrc/autograd/init.cpp:
```diff
@@ -1258,7 +1258,37 @@ PyObject* THPModule_increment_version(PyObject* _unused, PyObject* tensor) {
+    PyObject* _unused,
+    PyObject* tensor_or_list) {
+  HANDLE_TH_ERRORS
+  if (PyList_Check(tensor_or_list)) {
```
You most likely want to use sequence here https://docs.python.org/3/c-api/sequence.html so you don't have to do 2 passes to create the input in your AOTDispatch code above and can just pass the generator.

bdhirsh replied:
good point, will try it

bdhirsh replied:
noob q, should I be using an iterator instead of a sequence? https://docs.python.org/3/c-api/iter.html

I see that sequence exposes a `PySequence_GetItem` that maps to `seq[i]`, although I would imagine that if I'm passing in a generator then I would just want to be looping through the entire generator without using random access.

(e.g. in case, `gen = (args[i] for i in mutated_indices)`

bdhirsh committed:
Update on "torch.autograd.graph.increment_version: accept List[Tensor], use in AOTDispatcher"

The regression from https://github.com/pytorch/pytorch/issues/132281 pinpoints https://github.com/pytorch/pytorch/commit/e4ace1a396265706ddd0041bb6a26c18d2de1095 as the cause. The main delta that commit introduces is that we now manually check `is_inference()` and call `increment_version()` (a pybind call) on every mutated input tensor to the graph.

This PR attempts to reduce overhead a bit by bundling up all of those checks into a single pybind call, by:

(1) updating `torch.autograd.graph.increment_version()` to accept a `Union[Tensor, List[Tensor]]`

(2) updating its semantics to no-op if you pass in a tensor with no version counter, instead of erroring



[ghstack-poisoned]

albanD commented on torch/_C/__init__.pyi.in:
```diff
@@ -1331,8 +1331,7 @@ def autocast_increment_nesting() -> _int: ...
 def set_autocast_cache_enabled(enabled: _bool) -> None: ...
-def _increment_version(tensor: Tensor) -> None: ...
-def _increment_versions(tensors: List[Tensor]) -> None: ...
+def _increment_version(tensors: Iterable[Tensor]) -> None: ...
```
The type is not correct here. Does "Iterate-able" exists? :p

bdhirsh replied:
I think `Iterable` is actually the right type here? (not `Iterator` though)

Skylion007 replied:
https://docs.python.org/3/library/typing.html#typing.Iterable

Skylion007 replied:
```suggestion
def _increment_version(tensors: Union[Iterable[Tensor], Tensor]) -> None: ...
```

albanD commented on torch/autograd/graph.py:
```diff
@@ -215,20 +215,26 @@ def get_gradient_edge(tensor: torch.Tensor) -> GradientEdge:
+    we will not bump its version counter (because your tensor does not have one).
     """
-    torch._C._increment_version(tensor)
+    if isinstance(tensor, torch.Tensor):
+        torch._C._increment_version([tensor])
+    else:
+        torch._C._increment_version(tensor)
```
```suggestion
    if isinstance(tensor, torch.Tensor):
        tensor = (tensor,)
    torch._C._increment_version(tensor)
```

Not need to create a full list

albanD commented on torch/csrc/autograd/init.cpp:
```diff
@@ -1254,11 +1254,24 @@ static PyObject* len_torch_dispatch_stack(PyObject* _unused, PyObject* args) {
-      THPVariable_Check(tensor), "increment_version expect a Tensor as input");
-  torch::autograd::increment_version((THPVariable_Unpack(tensor)));
+      iterator != nullptr,
+      "increment_version expect a Tensor or Iterable[Tensor] as input");
```
At this point, it must be Iterate-able. Cannot be a Tensor !

albanD commented on torch/csrc/autograd/init.cpp:
```diff
@@ -1254,11 +1254,24 @@ static PyObject* len_torch_dispatch_stack(PyObject* _unused, PyObject* args) {
+    PyObject* _unused,
+    PyObject* tensor_list) {
   HANDLE_TH_ERRORS
+  PyObject* iterator = PyObject_GetIter(tensor_list);
```
Refcount for this and item below is missing

bdhirsh replied:
thanks you - as per offline, I'm now wrapping both `PyObject_GetIter` and `PyIter_Next` outputs in a `THPObjectPtr`, so they are properly decref'd on destruction.

bdhirsh committed:
Update on "torch.autograd.graph.increment_version: accept List[Tensor], use in AOTDispatcher"

The regression from https://github.com/pytorch/pytorch/issues/132281 pinpoints https://github.com/pytorch/pytorch/commit/e4ace1a396265706ddd0041bb6a26c18d2de1095 as the cause. The main delta that commit introduces is that we now manually check `is_inference()` and call `increment_version()` (a pybind call) on every mutated input tensor to the graph.

This PR attempts to reduce overhead a bit by bundling up all of those checks into a single pybind call, by:

(1) updating `torch.autograd.graph.increment_version()` to accept a `Union[Tensor, List[Tensor]]`

(2) updating its semantics to no-op if you pass in a tensor with no version counter, instead of erroring



[ghstack-poisoned]

Skylion007 commented on torch/autograd/graph.py:
```diff
@@ -215,19 +215,24 @@ def get_gradient_edge(tensor: torch.Tensor) -> GradientEdge:
     if you are doing inplace operation on the Tensor data in a way that Pytorch doesn't
     know about. For example a custom kernel that reads the Tensor data_ptr and modifies
-    the memory inplace based on this pointer.
+    the memory inplace based on this pointer.Can accept either a tensor, or a list of tensors.
```
Should have a space?

bdhirsh committed:
Update on "torch.autograd.graph.increment_version: accept List[Tensor], use in AOTDispatcher"

The regression from https://github.com/pytorch/pytorch/issues/132281 pinpoints https://github.com/pytorch/pytorch/commit/e4ace1a396265706ddd0041bb6a26c18d2de1095 as the cause. The main delta that commit introduces is that we now manually check `is_inference()` and call `increment_version()` (a pybind call) on every mutated input tensor to the graph.

This PR attempts to reduce overhead a bit by bundling up all of those checks into a single pybind call, by:

(1) updating `torch.autograd.graph.increment_version()` to accept a `Union[Tensor, List[Tensor]]`

(2) updating its semantics to no-op if you pass in a tensor with no version counter, instead of erroring



[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -3992,8 +3992,7 @@ def test_increment_version(self):
             a = torch.rand(5, requires_grad=True)
-        msg = "update to inference tensor outside InferenceMode"
-        with self.assertRaisesRegex(RuntimeError, msg):
+            # does not error
```
You need to fix indentation !
This is expected to check outside of inferenc emode !

bdhirsh replied:
pushed too quickly... fixing the indentation.

> This is expected to check outside of inferenc emode !

wdym? (now that we aren't raising an error when you call this fn on a tensor created in inference mode, I figured I would just make the test assert that we can call `increment_version` on a tensor created in inference mode without blowing up)

albanD reviewed (approved):
SGTM !

bdhirsh committed:
Update on "torch.autograd.graph.increment_version: accept List[Tensor], use in AOTDispatcher"

The regression from https://github.com/pytorch/pytorch/issues/132281 pinpoints https://github.com/pytorch/pytorch/commit/e4ace1a396265706ddd0041bb6a26c18d2de1095 as the cause. The main delta that commit introduces is that we now manually check `is_inference()` and call `increment_version()` (a pybind call) on every mutated input tensor to the graph.

This PR attempts to reduce overhead a bit by bundling up all of those checks into a single pybind call, by:

(1) updating `torch.autograd.graph.increment_version()` to accept a `Union[Tensor, List[Tensor]]`

(2) updating its semantics to no-op if you pass in a tensor with no version counter, instead of erroring



[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
